pycape/cape.py

```
@@ -6,20 +6,52 @@
 import pathlib
 import random
 import ssl
+from functools import wraps
 
 import websockets
 
 import serdio
 from pycape import attestation as attest
 from pycape import enclave_encrypt
 
+from .function_ref import FunctionRef
+
 _CAPE_CONFIG_PATH = pathlib.Path.home() / ".config" / "cape"
 _DISABLE_SSL = os.environ.get("CAPEDEV_DISABLE_SSL", False)
 
 logging.basicConfig(format="%(message)s")
 logger = logging.getLogger("pycape")
 
 
+def check_run(method):
+    """
+    check_run turns the positional arguments into function_ref if types match
+    """
+
+    @wraps(method)
+    def new_method(*args, **kwargs):
```

I still feel like writing this logic as a decorator is over-engineered. imo it would make sense if we were using this helper across a ton of other methods/functions, and wanted to eliminate boilerplate with it. but in this case, it's only used by one method, so it just ends up adding even more boilerplate. a simple helper function that sanitizes the inputs to `run` seems like the simpler choice

---

Are you saying to overload the function? This is to modify the inputs to the function to achieve
```
input = ""
f_id = "1s2tna3st5d"
f_hash = "s1th24s1t2h4s1th24"

result = cape.run(f_id, input, ...)  # implicit conversion into FunctionRef

f_ref = FunctionRef.from_id(f_id)  # construct FunctionRef from ID
result = cape.run(f_ref, input, ...)

f_ref_hashed = FunctionRef(function_id, function_hash)  # construct from ID + hash
result = cape.run(f_ref_hashed, input, ...)
```

Since we wanted to implicitly convert function_id

---

right, but I'm just saying that the implementation for this logic could simply be a helper function that `run` calls, instead of a fully-fledged decorator that wraps it. instead of 
```python
@check_run
def run(self, function_ref, ...):
```
i would suggest something like
```python
def run(self, function_ref, ...):
    function_ref = _convert_function_to_ref(function_ref)
```
the helper `_convert_function_to_ref` could also be used in `Cape.connect`, which I'm just now realizing probably shouldn't expect a different format of args than `Cape.run`

---

I can change the signature of Cape.connect to take function_ref. The decorator has the benefit of being able to check with args and kwargs at the same time. Specifically if the user passes in a non-keyed value then we can throw the error in the decorator.